### PR TITLE
feat(router): add support for Samsung Pay payment method

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This pr is to support for Samsung pay as a payment method.
This pr also adds a new field in merchant connector account create `connector_wallet_details` which is required to collect the required credentials for Samsung Pay. As this column in the data base stores the data in the encrypted format we also want to migrate apple pay certificates from metadata to `connector_wallet_details`, for which we are inserting apple pay details passed in the metadata to connector_wallet_details. 

Once the frontend changes are done to pass the apple pay credentials in the connector_wallet_details this fallback can be removed.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Test Samsung Pay flow
-> Enable Samsung Pay payment method for the connector
```
{
    "connector_type": "payment_processor",
    "connector_name": "cybersource",
    "connector_account_details": {
        "auth_type": "SignatureKey",
        "api_secret": "api_secret",
        "key1": "key1",
        "api_key": "api_key"
    },
    "test_mode": true,
    "disabled": false,
    "payment_methods_enabled": [
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "invoke_sdk_client",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "payment_experience": null,
                    "card_networks": [
                        "Visa",
                        "Mastercard",
                        "Discover",
                        "DinersClub"
                    ],
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "debit",
                    "payment_experience": null,
                    "card_networks": [
                        "Visa",
                        "Mastercard"
                    ],
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "bank_debit",
            "payment_method_types": [
                {
                    "payment_method_type": "ach",
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "minimum_amount": 0,
                    "maximum_amount": 10000
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "google_pay",
                    "payment_experience": "invoke_sdk_client",
                    "card_networks": null,
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "samsung_pay",
                    "payment_experience": "invoke_sdk_client",
                    "card_networks": null,
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "apple_pay",
                    "payment_experience": "invoke_sdk_client",
                    "card_networks": null,
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "paypal",
                    "payment_experience": "redirect_to_url",
                    "card_networks": null,
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
    "connector_wallets_details": {
        "samsung_pay": {
            "service_id": "49400558c67f4a97b3925f",
            "merchant_display_name": "Hyperswitch",
            "merchant_business_country": "IN",
            "allowed_brands": [
                "visa",
                "masterCard",
                "amex",
                "discover"
            ]
        }
    },
    "connector_webhook_details": {
        "merchant_secret": "merchant_secret"
    }
}
```
-> Create a payment
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_cjSDZPveYx7AbdI5cMRMuPKVdteJ7aIi3ULRasYhU89JtFRZWp9BeNn7PQDLQv8W' \
--data '{
  "amount": 100000,
  "currency": "INR",
  "capture_method": "automatic",
  "authentication_type": "no_three_ds",
  "confirm": false,
  "customer_id": "cu_1726743607"
}'
```
```
{
    "payment_id": "pay_on1wv5l8OM8A3VuDe2GN",
    "merchant_id": "merchant_1726725490",
    "status": "requires_payment_method",
    "amount": 100000,
    "net_amount": 100000,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": null,
    "client_secret": "pay_on1wv5l8OM8A3VuDe2GN_secret_o5JrhSWuWKjcUgrvASpi",
    "created": "2024-09-19T11:00:15.562Z",
    "currency": "INR",
    "customer_id": "cu_1726743615",
    "customer": {
        "id": "cu_1726743615",
        "name": null,
        "email": null,
        "phone": null,
        "phone_country_code": null
    },
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": null,
    "payment_method_data": null,
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": null,
    "name": null,
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": null,
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "cu_1726743615",
        "created_at": 1726743615,
        "expires": 1726747215,
        "secret": "epk_3d5ad79c88554ee7bf0748ae18ecd05b"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": null,
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_XdGC3A6KTKR5n2WDLON5",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": null,
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-09-19T11:15:15.562Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-09-19T11:00:15.680Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null,
    "order_tax_amount": null
}
```
-> Session call
```
curl --location 'http://localhost:8080/payments/session_tokens' \
--header 'Content-Type: application/json' \
--header 'x-browser-name: Safari' \
--header 'x-merchant-domain: sdk-test-app.netlify.app' \
--header 'api-key: pk_dev_c23e25c5054241288a778316e463e309' \
--data '{
    "payment_id": "pay_on1wv5l8OM8A3VuDe2GN",
    "wallets": [],
    "client_secret": "pay_on1wv5l8OM8A3VuDe2GN_secret_o5JrhSWuWKjcUgrvASpi"
}'
```
```
{
    "payment_id": "pay_on1wv5l8OM8A3VuDe2GN",
    "client_secret": "pay_on1wv5l8OM8A3VuDe2GN_secret_o5JrhSWuWKjcUgrvASpi",
    "session_token": [
        {
            "wallet_name": "samsung_pay",
            "version": "2",
            "service_id": "49400558c67f4a97b3925f",
            "order_number": "pay_on1wv5l8OM8A3VuDe2GN",
            "merchant": {
                "name": "Hyperswitch",
                "url": "sdk-test-app.netlify.app",
                "country_code": "IN"
            },
            "amount": {
                "option": "FORMAT_TOTAL_PRICE_ONLY",
                "currency_code": "INR",
                "total": "1000.00"
            },
            "protocol": "PROTOCOL3DS",
            "allowed_brands": [
                "visa",
                "masterCard",
                "amex",
                "discover"
            ]
        }
    ]
}
```
-> Confirm call
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_cjSDZPveYx7AbdI5cMRMuPKVdteJ7aIi3ULRasYhU89JtFRZWp9BeNn7PQDLQv8W' \
--data-raw '{
    "amount": 1,
    "currency": "USD",
    "confirm": true,
    "business_country": "US",
    "business_label": "default",
    "amount_to_capture": 1,
    "customer_id": "custhype123",
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "return_url": "https://google.com",
    "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "payment_method": "wallet",
    "payment_method_type": "samsung_pay",
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "payment_method_data": {
        "wallet": {
            "samsung_pay": {
                "payment_credential": {
                    "card_brand": "visa",
                    "recurring_payment": false,
                    "card_last4digits": "1234",
                    "method": "3DS",
                    "3_d_s": {
                        "type": "S",
                        "version": "100",
                        "data": "encrypted samsung pay data"
                    }
                }
            }
        }
    }
}'
```
```
{
    "payment_id": "pay_0lHEl7IAq7cFn680x4HJ",
    "merchant_id": "merchant_1726725490",
    "status": "failed",
    "amount": 1,
    "net_amount": 1,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": "cybersource",
    "client_secret": "pay_0lHEl7IAq7cFn680x4HJ_secret_3bsNpl8N5rArpWizVWik",
    "created": "2024-09-19T11:28:58.596Z",
    "currency": "USD",
    "customer_id": "custhype123",
    "customer": {
        "id": "custhype123",
        "name": "Joseph Doe",
        "email": "something@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "wallet",
    "payment_method_data": {
        "wallet": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": null,
    "email": "something@gmail.com",
    "name": "Joseph Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "Juspay",
    "statement_descriptor_suffix": "Router",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": "INVALID_DATA",
    "error_message": "Declined - One or more fields in the request contains invalid data, detailed_error_information: paymentInformation.fluidData.value : INVALID_DATA",
    "unified_code": "UE_000",
    "unified_message": "Something went wrong",
    "payment_experience": null,
    "payment_method_type": "samsung_pay",
    "connector_label": "cybersource_US_default_default",
    "business_country": "US",
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "custhype123",
        "created_at": 1726745338,
        "expires": 1726748938,
        "secret": "epk_48f2e9de3ad641578f3df9e9a46d9c80"
    },
    "manual_retry_allowed": true,
    "connector_transaction_id": null,
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_XdGC3A6KTKR5n2WDLON5",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_TO5NdOnj3ohWTfxBKGsc",
    "incremental_authorization_allowed": false,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-09-19T11:43:58.596Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-09-19T11:29:00.222Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null,
    "order_tax_amount": null
}
```
The status will be failed as Samsung does not have a test card and cybersource returns invalid data error for live cards  

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: config/development.toml [hunk 1]
```diff
@@ -447,6 +447,7 @@ credit = { currency = "USD,GBP,EUR" }
 debit = { currency = "USD,GBP,EUR" }
 apple_pay = { currency = "USD,GBP,EUR" }
 google_pay = { currency = "USD,GBP,EUR" }
+samsung_pay = { currency = "USD,GBP,EUR" }
```
Comment (SanchithHegde) [inline config/development.toml:450]: Can we update this for the other config files as well?

File: crates/api_models/src/admin.rs [hunk 2]
```diff
@@ -1503,6 +1527,28 @@ pub struct MerchantConnectorUpdate {
...
+
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
+#[serde(deny_unknown_fields)]
+
+pub struct ConnectorWalletDetails {
+    /// This field contains the Apple Pay certificates and credentials for iOS and Web Apple Pay flow
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[schema(value_type = Option<Object>)]
+    pub apple_pay_combined: Option<pii::SecretSerdeValue>,
+    /// This field is for our legacy that contains the Apple Pay certificates and credentials for only iOS Apple Pay flow
```
Comment (SanchithHegde) [inline crates/api_models/src/admin.rs:1544]: I think we're missing a word here, "legacy ..."?

File: crates/api_models/src/payments.rs [hunk 3]
```diff
@@ -2787,9 +2787,36 @@ impl GetAddressFromPaymentMethodData for WalletData {
...
 #[serde(rename_all = "snake_case")]
 pub struct SamsungPayWalletData {
-    /// The encrypted payment token from Samsung
+    pub payment_credential: SamsungPayWalletCredentials,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, serde::Deserialize, serde::Serialize, ToSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct SamsungPayWalletCredentials {
+    pub method: Option<String>,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:2796]: No documentation comments for `method`?

File: crates/api_models/src/payments.rs [hunk 4]
```diff
@@ -4747,6 +4774,20 @@ pub struct GpaySessionTokenData {
     pub data: GpayMetaData,
 }
 
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct SamsungSessionTokenData {
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:4778]: Nit: Naming.

```suggestion
pub struct SamsungPaySessionTokenData {
```

File: crates/api_models/src/payments.rs [hunk 5]
```diff
@@ -4747,6 +4774,20 @@ pub struct GpaySessionTokenData {
...
+    #[serde(rename = "samsung_pay")]
+    pub data: SamsungMetaData,
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct SamsungMetaData {
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:4784]:
```suggestion
pub struct SamsungPayMetadata {
```

File: crates/api_models/src/payments.rs [hunk 6]
```diff
@@ -4747,6 +4774,20 @@ pub struct GpaySessionTokenData {
...
+    pub service_id: String,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:4785]: We can wrap this within `Secret` if this is considered sensitive information.

File: crates/api_models/src/payments.rs [hunk 7]
```diff
@@ -4969,6 +5012,68 @@ pub struct GooglePaySessionResponse {
     pub secrets: Option<SecretInfoToInitiateSdk>,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, ToSchema)]
+#[serde(rename_all = "lowercase")]
+pub struct SamsungPaySessionTokenResponse {
+    /// Samsung Pay API version
+    pub version: String,
+    /// Samsung Pay service ID to which session call needs to be made
+    pub service_id: String,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:5021]: Same here.

File: crates/api_models/src/payments.rs [hunk 8]
```diff
@@ -4969,6 +5012,68 @@ pub struct GooglePaySessionResponse {
...
+    pub amount: AmountDetails,
+    /// Payment protocol type
+    pub protocol: ProtocolType,
+    /// List of supported card brands
+    pub allowed_brands: Vec<String>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, ToSchema)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum ProtocolType {
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:5037]: Should we consider calling this `SamsungPayProtocolType`?

File: crates/api_models/src/payments.rs [hunk 9]
```diff
@@ -4969,6 +5012,68 @@ pub struct GooglePaySessionResponse {
...
+    pub service_id: String,
+    /// Order number of the transaction
+    pub order_number: String,
+    /// Filed containing merchant information
+    #[serde(rename = "merchant")]
+    pub merchant_payment_information: MerchantPaymentInformation,
+    /// Filed containing the payment amount
+    pub amount: AmountDetails,
+    /// Payment protocol type
+    pub protocol: ProtocolType,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:5030]: Same for `MerchantPaymentInformation`, `AmountDetails` and `AmountFormat`, aren't they going to be Samsung Pay specific?

File: crates/hyperswitch_domain_models/src/payment_method_data.rs [hunk 10]
```diff
@@ -135,10 +135,37 @@ pub struct MifinityData {
...
-    /// The encrypted payment token from Samsung
-    pub token: Secret<String>,
+    pub payment_credential: SamsungPayWalletCredentials,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct SamsungPayWalletCredentials {
+    pub method: Option<String>,
+    /// Value if credential is enabled for recurring payment
```
Comment (SanchithHegde) [inline crates/hyperswitch_domain_models/src/payment_method_data.rs:147]: You can ignore doc comments for domain models, if we're not gonna bother maintaining them and keeping them up-to-date.

File: crates/router/src/connector/cybersource/transformers.rs [hunk 11]
```diff
@@ -436,12 +436,28 @@ pub struct FluidData {
 
 pub const FLUID_DATA_DESCRIPTOR: &str = "RklEPUNPTU1PTi5BUFBMRS5JTkFQUC5QQVlNRU5U";
 
+pub const FLUID_DATA_DESCRIPTOR_FOR_SAMSUNG_PAY: &str =
+    "RklEPUNPTU1PTi5TQU1TVU5HLklOQVBQLlBBWU1FTlQ=";
```
Comment (SanchithHegde) [inline crates/router/src/connector/cybersource/transformers.rs:440]: Rather, keep the plaintext value `FID=COMMON.SAMSUNG.INAPP.PAYMENT` here in the constant, then base64 encode it in code.

File: crates/router/src/types/transformers.rs [hunk 12]
```diff
@@ -1068,6 +1068,18 @@ impl ForeignTryFrom<domain::MerchantConnectorAccount>
...
                 .transpose()?
                 .map(api_models::admin::AdditionalMerchantData::foreign_from),
+            connector_wallets_details: item
+                .connector_wallets_details
+                .map(|data| {
+                    let data = data.into_inner();
+                    serde_json::Value::parse_value::<api_models::admin::ConnectorWalletDetails>(
+                        data.expose(),
+                        "ConnectorWalletDetails",
+                    )
```
Comment (SanchithHegde) [inline crates/router/src/types/transformers.rs:1078]: Can't we call `data.expose().parse_value::<api_models::admin::ConnectorWalletDetails>("ConnectorWalletDetails")`?

File: crates/api_models/src/payments.rs (hunk 6 above)
Comment (ShankarSinghC) [reply crates/api_models/src/payments.rs:4786]: > We can wrap this within `Secret` if this is considered sensitive information.

No, it is not a secret value. We send this is session token response to sdk in order to make call to Samsung Pay.

File: crates/router/src/types/transformers.rs (hunk 12 above)
Comment (SanchithHegde) [reply crates/router/src/types/transformers.rs]: Can you make the same change elsewhere in the file?

File: crates/router/src/connector/cybersource/transformers.rs [hunk 13]
```diff
@@ -504,19 +520,23 @@ pub struct AdditionalAmount {
...
     GooglePay,
+    SamsungPay,
 }
 
 #[derive(Debug, Serialize)]
 pub enum TransactionType {
     #[serde(rename = "1")]
     ApplePay,
+    #[serde(rename = "1")]
+    SamsungPay,
```
[the remark this answers is not in the document]
Comment (ShankarSinghC) [reply crates/router/src/connector/cybersource/transformers.rs:531]: Yes, it is mentioned in the cybersource documentation.

**Review by SanchithHegde — APPROVED** [review]

**Review by deepanshu-iiitu — DISMISSED** [review]

**Review by deepanshu-iiitu — COMMENTED** [review]
Approving on behalf of connectors

**Review by Chethan-rao — DISMISSED** [review]

**Review by apoorvdixit88 — APPROVED** [review]
Dashboard specific change looks fine

**Review by Chethan-rao — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]

**Pull request merged**